Library/Homebrew/cmd/test-bot.rb

```diff
@@ -944,11 +953,21 @@ def test_bot
 
           if step.has_output?
             # Remove invalid XML CData characters from step output.
-            output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            valid_xml_chars = ""
+            if ruby_has_encoding?
+              # This is the regex for valid XML chars, but only works in Ruby 2.0+
+              # /[]\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/
+              # For 1.9 compatibility, use the inverse of that, which staus under \u10000
+              invalid_xml_pat = /[\x00-\x08\x0B\x0C\x0E-\x1F\uD800-\uDFFF\uFFFE\uFFFF]/
+              output = step.output.gsub(invalid_xml_pat, "\uFFFD")
+            else
+              output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            end
 
             if output.bytesize > BYTES_IN_1_MEGABYTE
-              output = "truncated output to 1MB:\n" \
-                + output.slice(-BYTES_IN_1_MEGABYTE, BYTES_IN_1_MEGABYTE)
+              slice_start = [0, -BYTES_IN_1_MEGABYTE].max
```

Comment:
No, because the code is calculating the slice start and end points in bytes, but the arguments to `slice()` are characters. My adjustment here is to account for that bytesize vs character differences and prevent overruns.

And the fact that `slice()` takes characters prevents it from cutting in to the middle of a UTF-8 byte sequence. (It doesn't account for surrogate pairs, though. That's an edge case, even in terms of validity, and I doubt we'd ever see it.)

Grapheme clusters: yeah, but we're already truncating at an arbitrary point; it's probably going to be in the middle of a word, so the stuff right around the cutoff is probably going to be junk to a human anyway.